### PR TITLE
CMPT-33662 resource secgroup, resource secgroup_rule

### PR DESCRIPTION
**Pull request by volosatik**
CMPT-33662
CMPT-33663

### CODE REVIEW

File: examples/firewall/networking_secgroup/main.tf [hunk 1]
```diff
@@ -0,0 +1,4 @@
+resource "vkcs_networking_secgroup" "custom_secgroup" {
```
Comment (ftersin) [inline examples/firewall/networking_secgroup/main.tf:1]: use a group from root firewall example

File: examples/firewall/networking_secgroup/main-datasource.tf [hunk 2]
```diff
@@ -0,0 +1,6 @@
+data "vkcs_networking_secgroup" "another-secgroup" {
```
Comment (ftersin) [inline examples/firewall/networking_secgroup/main-datasource.tf:1]: give the same name as for the the group

File: examples/firewall/networking_secgroup_rule/main-one-port.tf [hunk 3]
```diff
@@ -0,0 +1,8 @@
+resource "vkcs_networking_secgroup_rule" "in_one_tcp" {
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-one-port.tf:1]: copy this group from root firewall example

File: examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf [hunk 4]
```diff
@@ -0,0 +1,6 @@
+resource "vkcs_networking_secgroup_rule" "in_all_udp" {
+  direction         = "ingress"
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf:2]: lets add descriptions for all rules

File: examples/firewall/networking_secgroup_rule/main.tf [hunk 5]
```diff
@@ -0,0 +1,5 @@
+resource "vkcs_networking_secgroup_rule" "out_all_protocols" {
+  direction         = "egress"
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main.tf:2]: add a comment why we need egress rule

File: templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl [hunk 6]
```diff
@@ -13,6 +13,12 @@ description: |-
 ## Example Usage
 {{tffile .ExampleFile}}
 
+## Rule for all port udp protocol
+{{tffile "examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf"}}
+
+## Rule for one port tcp protocol
```
Comment (ftersin) [inline templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl:19]: i believe this must be the first example

File: examples/firewall/networking_secgroup_rule/main-one-port.tf [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+  direction         = "ingress"
+  protocol          = "tcp"
+  port_range_min    = 11435
+  port_range_max    = 11435
+  remote_ip_prefix  = vkcs_networking_subnet.db.cidr
+  security_group_id = vkcs_networking_secgroup.custom_secgroup.id
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-one-port.tf:7]: why do we have both remote_ip_prefix and security_group_id here?

File: examples/firewall/networking_secgroup_rule/main.tf [hunk 8]
```diff
@@ -0,0 +1,5 @@
...
+  remote_ip_prefix  = "10.10.0.0/16"
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main.tf:3]: let it be all address space

File: examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf [hunk 9]
```diff
@@ -0,0 +1,6 @@
...
+  protocol          = "udp"
+  remote_ip_prefix  = vkcs_networking_subnet.db.cidr
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf:4]: use remote_security_group for this rule

File: examples/firewall/networking_secgroup/main.tf (hunk 1 above)
Comment (volosatik) [reply examples/firewall/networking_secgroup/main.tf]: used etcd

File: examples/firewall/networking_secgroup_rule/main-one-port.tf (hunk 3 above)
Comment (volosatik) [reply examples/firewall/networking_secgroup_rule/main-one-port.tf]: used etcd

File: examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf (hunk 4 above)
Comment (volosatik) [reply examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf:3]: added

File: examples/firewall/networking_secgroup_rule/main.tf (hunk 5 above)
Comment (volosatik) [reply examples/firewall/networking_secgroup_rule/main.tf:4]: added

File: templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl (hunk 6 above)
Comment (volosatik) [reply templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl:19]: changed

File: examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf (hunk 9 above)
Comment (volosatik) [reply examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf]: changed

File: examples/firewall/networking_secgroup_rule/main.tf (hunk 8 above)
Comment (volosatik) [reply examples/firewall/networking_secgroup_rule/main.tf]: changed

File: examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf (hunk 4 above)
Comment (paaanic) [reply examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf:3]: typo: portocol -> protocol

File: examples/firewall/networking_secgroup_rule/main.tf [hunk 10]
```diff
@@ -0,0 +1,9 @@
+resource "vkcs_networking_secgroup_rule" "etcd_app_clients" {
+  description       = "etcd app clients rule - allow one port for one porotocol tcp"
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main.tf:2]: get the description from root example

File: templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl [hunk 11]
```diff
@@ -10,9 +10,15 @@ description: |-
 
 {{ .Description }}
 
-## Example Usage
+## Example Usage one protocol one port
```
Comment (ftersin) [inline templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl:13]: "Example Usage" is a chapter name. Add example names on the next level. See compute examples

Comment (ftersin) [reply templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl:13]: Name it "Basic rule"

File: templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl [hunk 12]
```diff
@@ -10,9 +10,15 @@ description: |-
...
 {{tffile .ExampleFile}}
 
+## Rule for all port udp protocol
```
Comment (ftersin) [inline templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl:16]: Rule for all ports

File: templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl [hunk 13]
```diff
@@ -10,9 +10,15 @@ description: |-
...
+{{tffile "examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf"}}
+
+## Rule for all ports all protocols
```
Comment (ftersin) [inline templates/firewall/resources/vkcs_networking_secgroup_rule.md.tmpl:19]: Rule for all protocols

File: examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf [hunk 14]
```diff
@@ -0,0 +1,7 @@
+resource "vkcs_networking_secgroup_rule" "in_all_udp" {
+  description       = "allow all ports for one porotocol udp"
+  direction         = "ingress"
+  protocol          = "udp"
+  remote_group_id   = vkcs_networking_secgroup.etcd.id
+  security_group_id = vkcs_networking_secgroup.etcd.id
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf:6]: Place it right after description

File: examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf [hunk 15]
```diff
@@ -0,0 +1,7 @@
+resource "vkcs_networking_secgroup_rule" "in_all_protocols" {
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf:1]: all_protocols

File: examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf [hunk 16]
```diff
@@ -0,0 +1,7 @@
...
+  description       = "allow all ports for all protocols"
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf:2]: Any inbound traffic

File: examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf [hunk 17]
```diff
@@ -0,0 +1,7 @@
...
+  # Allow access to any sources from our machines
+  direction         = "ingress"
+  remote_ip_prefix  = "0.0.0.0/0"
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf:5]: use remote_group_id instead

File: examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf [hunk 18]
```diff
@@ -0,0 +1,7 @@
+resource "vkcs_networking_secgroup_rule" "in_all_udp" {
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf:1]: all_udp

File: examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf [hunk 19]
```diff
@@ -0,0 +1,7 @@
...
+  description       = "allow all ports for one porotocol udp"
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-protocol.tf:2]: All inbound UDP traffic from etcd hosts

File: examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf [hunk 20]
```diff
@@ -0,0 +1,7 @@
+resource "vkcs_networking_secgroup_rule" "all" {
+  description       = "Any inbound traffic from etcd hosts"
+  security_group_id = vkcs_networking_secgroup.etcd.id
+  # Allow access to any sources from our machines
```
Comment (ftersin) [inline examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf:4]: remove this comment

Comment (volosatik) [reply examples/firewall/networking_secgroup_rule/main-all-ports-all-protocol.tf]: removed

**Review by ftersin — APPROVED** [review]

**Pull request merged**